neqo-transport/src/streams.rs

```diff
@@ -0,0 +1,464 @@
+
+            self.remote_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+            self.remote_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+        }
+
+        self.send_streams
+            .write_frames(priority, builder, tokens, stats)
+    }
+
+    pub fn lost(&mut self, token: &RecoveryToken) {
+        match token {
+            RecoveryToken::Stream(st) => self.send_streams.lost(&st),
+            RecoveryToken::ResetStream { stream_id } => self.send_streams.reset_lost(*stream_id),
+            RecoveryToken::StreamDataBlocked { stream_id, limit } => {
+                self.send_streams.blocked_lost(*stream_id, *limit)
+            }
+            RecoveryToken::MaxStreamData {
+                stream_id,
+                max_data,
+            } => {
+                if let Ok((_, Some(rs))) = self.obtain_stream(*stream_id) {
+                    rs.max_stream_data_lost(*max_data);
+                }
+            }
+            RecoveryToken::StopSending { stream_id } => {
+                if let Ok((_, Some(rs))) = self.obtain_stream(*stream_id) {
+                    rs.stop_sending_lost();
+                }
+            }
+            RecoveryToken::StreamsBlocked { stream_type, limit } => {
+                self.local_streams_fc[*stream_type].lost(*limit);
+            }
+            RecoveryToken::MaxStreams {
+                stream_type,
+                max_streams,
+            } => {
+                self.remote_streams_fc[*stream_type].lost(*max_streams);
+            }
+            RecoveryToken::DataBlocked(limit) => self.flow_control_sender.borrow_mut().lost(*limit),
+            RecoveryToken::MaxData(maximum_data) => {
+                self.flow_control_receiver.borrow_mut().lost(*maximum_data)
+            }
+            _ => unreachable!("This is not a stream RecoveryToken"),
+        }
+    }
+
+    pub fn acked(&mut self, token: &RecoveryToken) {
+        match token {
+            RecoveryToken::Stream(st) => self.send_streams.acked(st),
+            RecoveryToken::ResetStream { stream_id } => self.send_streams.reset_acked(*stream_id),
+            RecoveryToken::StopSending { stream_id } => {
+                if let Ok((_, Some(rs))) = self.obtain_stream(*stream_id) {
+                    rs.stop_sending_acked();
+                }
+            }
+            _ => unreachable!("This is not a stream RecoveryToken"),
```

Comment:
Should this be more tolerant of tokens that *are* for streams?  Like DataBlocked and Maxdata?

One way to avoid special matching logic is to have a method on RecoveryToken called `is_stream()` that the connection uses to route to here.  That saves us from having to replicate the matching logic in connection for stream-related tokens.

Reply:
I was thinking about adding something like is_stream() to RecoveryToken, but decided to leave it for a followup.